## **WORKSHOP 001 - NOTEBOOK #2: Exploratory Data Analysis (EDA)**

This notebook focuses on conducting **Exploratory Data Analysis (EDA)** on the candidates' dataset. EDA is a fundamental step in the data analysis process, as it allows us to understand the structure and characteristics of the data. By examining the dataset, we can identify patterns, relationships, and key insights that inform further analysis and decision-making.

In this notebook, we will explore the dataset using a range of statistical and visual techniques. We will assess the distribution of variables and investigate correlations between them. Through this process, we aim to develop a thorough understanding of the dataset and extract valuable insights.

---

### **Setting Environment**

In [1]:
import os 
print(os.getcwd())

try:
    os.chdir("../../workshop-001")

except FileNotFoundError:
    print("""
        FileNotFoundError - The directory may not exist or you might not be in the specified path.
        If this has already worked, do not run this block again, as the current directory is already set to workshop-001.
        """)
    
print(os.getcwd())

d:\U\FIFTH SEMESTER\ETL\workshop-001\notebooks
d:\U\FIFTH SEMESTER\ETL\workshop-001


## **Load Data**

### **Import dependencies**

We will use Pandas to analyse the data within the DataFrame, while Matplotlib and Seaborn will be employed to generate graphs that illustrate the insights from the database.

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
plt.style.use('ggplot')

In [4]:
from functions.db_connection.connection import creating_engine

### **Create engine**

The connection process has been assigned to a Python script called _connection.py_, where the function _creating_engine_ is responsible for setting up the connection to a PostgreSQL database using SQLAlchemy.

In [6]:
engine = creating_engine()

### **Load Database**

We retrieve the dataset from a table in a PostgreSQL database connected using the SQLAlchemy engine.

In [7]:
df = pd.read_sql_table("candidates_raw", engine, parse_dates=["Application Date"])
df

First Name   Last Name                       Email Application Date  \
0      Bernadette   Langworth         leonard91@yahoo.com       2021-02-26   
1          Camryn    Reynolds         zelda56@hotmail.com       2021-09-09   
2           Larue      Spinka    okey_schultz41@gmail.com       2020-04-14   
3            Arch      Spinka      elvera_kulas@yahoo.com       2020-10-01   
4           Larue  Altenwerth   minnie.gislason@gmail.com       2020-05-20   
...           ...         ...                         ...              ...   
49995     Bethany     Shields  rocky_mitchell@hotmail.com       2022-01-09   
49996         Era  Swaniawski    dolores.roob@hotmail.com       2020-06-02   
49997      Martin       Lakin   savanah.stracke@gmail.com       2018-12-15   
49998       Aliya   Abernathy  vivienne.fritsch@yahoo.com       2020-05-30   
49999     Coleman      Wisozk   abigayle.crooks@yahoo.com       2022-06-13   

                  Country  YOE  Seniority                         Technology  \
0                  Norway    2     Intern                      Data Engineer   
1                  Panama   10     Intern                      Data Engineer   
2                 Belarus    4  Mid-Level                     Client Success   
3                 Eritrea   25    Trainee                          QA Manual   
4                 Myanmar   13  Mid-Level  Social Media Community Management   
...                   ...  ...        ...                                ...   
49995  Dominican Republic   27    Trainee                           Security   
49996             Morocco   21       Lead                   Game Development   
49997              Uganda   20    Trainee              System Administration   
49998      Czech Republic   20     Senior            Database Administration   
49999               Palau   15     Intern                           Mulesoft   

       Code Challenge Score  Technical Interview Score  
0                         3                          3  
1                         2                         10  
2                        10                          9  
3                         7                          1  
4                         9                          7  
...                     ...                        ...  
49995                     2                          1  
49996                     1                          2  
49997                     6                          1  
49998                     0                          0  
49999                     3                          1  

[50000 rows x 10 columns]

### **Data Preprocessing**

The column names have been converted to [snake_case](https://www.theserverside.com/definition/Snake-case), which simplifies using df.query() and makes the process easier.

In [9]:
renamed_columns = {
    'First Name': 'first_name',
    'Last Name': 'last_name',
    'Email': 'email',
    'Application Date': 'application_date',
    'Country': 'country',
    'YOE': 'yoe',
    'Seniority': 'seniority',
    'Technology': 'technology',
    'Code Challenge Score': 'code_challenge_score',
    'Technical Interview Score': 'technical_interview_score'
}

df = df.rename(columns=renamed_columns)

### **Understanding the Dataset**

#### **Data column details**

The output of `df.info()` provides a summary of our dataset. 

It shows that the DataFrame contains 50,000 entries and 10 columns, with each column having *no missing values*. 

The data types are appropriately assigned: six columns are treated as objects, one is a datetime, and three are integers. 

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   first_name                 50000 non-null  object        
 1   last_name                  50000 non-null  object        
 2   email                      50000 non-null  object        
 3   application_date           50000 non-null  datetime64[ns]
 4   country                    50000 non-null  object        
 5   yoe                        50000 non-null  int64         
 6   seniority                  50000 non-null  object        
 7   technology                 50000 non-null  object        
 8   code_challenge_score       50000 non-null  int64         
 9   technical_interview_score  50000 non-null  int64         
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 3.8+ MB


#### **Examining the dataset for duplicate rows.**

In [12]:
df.duplicated().sum()

np.int64(0)

The result shows that there are *no duplicate rows* in the dataset. This means every row in the dataset is unique.

#### **Examining the dataset for null values.**

In [14]:
df.isnull().sum()

first_name                   0
last_name                    0
email                        0
application_date             0
country                      0
yoe                          0
seniority                    0
technology                   0
code_challenge_score         0
technical_interview_score    0
dtype: int64

This result shows that every column has 0 missing values, indicating that there is no missing data in any of them.

#### **Counting unique values in each column**

In [16]:
df.nunique()

first_name                    3007
last_name                      474
email                        49833
application_date              1646
country                        244
yoe                             31
seniority                        7
technology                      24
code_challenge_score            11
technical_interview_score       11
dtype: int64

This output displays the number of unique values in each column for the dataset containing 50,000 entries. 

For example, almost every email is unique, while columns such as seniority and scores show a limited range of distinct values. 

Considering the overall number of records, it is unusual to observe a number of repeated emails; indeed, the difference between the total records and the email count suggests that _approximately 167 individuals may have reapplied_.

**Analising those duplicated records**

In [17]:
duplicated_emails = df.loc[df.duplicated(subset=['email'], keep=False)]
duplicated_emails["email"].value_counts()

email
marianne31@yahoo.com        3
fern70@gmail.com            3
sandra83@gmail.com          2
dewayne50@gmail.com         2
matilda17@gmail.com         2
                           ..
marjolaine91@hotmail.com    2
jazmin54@gmail.com          2
reyna2@hotmail.com          2
kasandra68@hotmail.com      2
easter75@gmail.com          2
Name: count, Length: 165, dtype: int64

This result shows that there are 165 repeated email addresses in the dataset, with two addresses appearing three times, indicating that some individuals may have reapplied.

#### **Data Analysis**

In [18]:
df.describe()

application_date           yoe  code_challenge_score  \
count                       50000  50000.000000          50000.000000   
mean   2020-04-03 23:04:14.592000     15.286980              4.996400   
min           2018-01-01 00:00:00      0.000000              0.000000   
25%           2019-02-17 00:00:00      8.000000              2.000000   
50%           2020-04-06 00:00:00     15.000000              5.000000   
75%           2021-05-21 00:00:00     23.000000              8.000000   
max           2022-07-04 00:00:00     30.000000             10.000000   
std                           NaN      8.830652              3.166896   

       technical_interview_score  
count               50000.000000  
mean                    5.003880  
min                     0.000000  
25%                     2.000000  
50%                     5.000000  
75%                     8.000000  
max                    10.000000  
std                     3.165082

The dataset covers application dates from 1 January 2018 to 4 July 2022, which explains the limited data for 2022. 

The average years of experience is 15.29, although the high standard deviation indicates significant variability among candidates. 

Additionally, both the code challenge and technical interview scores average around 5, suggesting that fewer than half of the applicants are likely to be accepted.

#### **Analising Code Challenge and Technical Interview Scores**

In [21]:
score_count = (df[['code_challenge_score', 'technical_interview_score']].agg(pd.Series.value_counts)
                .sort_index()
                .rename_axis('score'))
score_count

code_challenge_score  technical_interview_score
score                                                 
0                      4502                       4539
1                      4590                       4588
2                      4579                       4500
3                      4678                       4528
4                      4521                       4578
5                      4479                       4527
6                      4419                       4471
7                      4506                       4578
8                      4619                       4555
9                      4519                       4597
10                     4588                       4539

In [22]:
total_applications = len(df)

code_challenge_percentage = (df['code_challenge_score'] >= 7).mean() * 100

technical_interview_percentage = (df['technical_interview_score'] >= 7).mean() * 100

hired_percentage = len(df[(df['code_challenge_score'] >= 7) & (df['technical_interview_score'] >= 7)]) / total_applications * 100

print("Percentage with Code Challenge Score >=7: {:.2f}%".format(code_challenge_percentage))
print("Percentage with Technical Interview Score >=7: {:.2f}%".format(technical_interview_percentage))
print("Percentage meeting both criteria (hired): {:.2f}%".format(hired_percentage))


Percentage with Code Challenge Score >=7: 36.46%
Percentage with Technical Interview Score >=7: 36.54%
Percentage meeting both criteria (hired): 13.40%


Based on the results, about 36% of candidates score 7 or above on each test. 

Given the requirement to score at least 7 on both assessments, roughly 13% of applicants would meet the criteria, indicating a selective hiring process.

The similar counts across the score range suggest that the scores are uniformly distributed.